vortex-scalar/src/decimal/mod.rs

```diff
@@ -321,8 +157,125 @@ impl<'a> DecimalScalar<'a> {
     }
 
     /// Returns the decimal value, or None if null.
-    pub fn decimal_value(&self) -> &Option<DecimalValue> {
-        &self.value
+    pub fn decimal_value(&self) -> Option<DecimalValue> {
+        self.value
+    }
+
+    /// Cast decimal scalar to another data type.
+    pub(crate) fn cast(&self, dtype: &DType) -> VortexResult<Scalar> {
+        match dtype {
+            DType::Decimal(target_dtype, target_nullability) => {
+                // Cast between decimal types
+                if self.decimal_type == *target_dtype {
+                    // Same decimal type, just change nullability if needed
+                    return Ok(Scalar::new(
+                        dtype.clone(),
+                        ScalarValue(InnerScalarValue::Decimal(
+                            self.value.unwrap_or(DecimalValue::I128(0)),
+                        )),
+                    ));
+                }
+
+                // Different precision/scale - need to implement scaling logic
+                // For now, we'll do a simple value preservation without scaling
+                // TODO: Implement proper decimal scaling logic
+                if let Some(value) = &self.value {
+                    Ok(Scalar::decimal(*value, *target_dtype, *target_nullability))
```

Comment:
should we just fail in this case instead? when do we actually cast decimals

Reply:
now i'm actually curious what duckdb does for casting one decimal to another precision